Lib/asyncio/unix_events.py

```diff
@@ -197,30 +197,31 @@ async def _make_subprocess_transport(self, protocol, args, shell,
                                          extra=None, **kwargs):
         with warnings.catch_warnings():
             warnings.simplefilter('ignore', DeprecationWarning)
-            with events.get_child_watcher() as watcher:
-                if not watcher.is_active():
-                    # Check early.
-                    # Raising exception before process creation
-                    # prevents subprocess execution if the watcher
-                    # is not ready to handle it.
-                    raise RuntimeError("asyncio.get_child_watcher() is not activated, "
-                                    "subprocess support is not installed.")
-                waiter = self.create_future()
-                transp = _UnixSubprocessTransport(self, protocol, args, shell,
-                                                stdin, stdout, stderr, bufsize,
-                                                waiter=waiter, extra=extra,
-                                                **kwargs)
-
-                watcher.add_child_handler(transp.get_pid(),
-                                        self._child_watcher_callback, transp)
-                try:
-                    await waiter
-                except (SystemExit, KeyboardInterrupt):
-                    raise
-                except BaseException:
-                    transp.close()
-                    await transp._wait()
-                    raise
+            watcher = events.get_child_watcher()
```

Comment:
I just learned that warnings.catch_warnings() is not threadsafe, so this solution still isn't 100% safe -- if another thread happens to be running while this code is active it may also get its warnings suppressed. It's just harder to reproduce because this code executes for only a fraction of a msec.

I'm not sure what to do other than introducing a private API with a different name and trying to call that, falling back on the public API only if it doesn't exist. This could still be wrong if someone subclasses a standard policy class and overrides get_child_watcher -- to be absolutely sure we might have to check the class and check that its get_child_watcher is the standard one.

Or we'll have to roll back the deprecation warning on get_child_watcher(). That actually sounds fairly plausible. @graingert: What do you think of that?

Reply:
You could introduce a thread local flag that prevents the warning being issued

Reply:
I like the thread local. @kumaraditya303 ?

Reply:
> I just learned that warnings.catch_warnings() is not threadsafe, so this solution still isn't 100% safe -- if another thread happens to be running while this code is active it may also get its warnings suppressed. It's just harder to reproduce because this code executes for only a fraction of a msec.

I don't understand the problem we are solving here TBH. Sure, warnings module is not thread safe but that doesn't mean we need to try hard to make it thread safe. There are plenty of occurrences of ignoring warnings in the stdlib. [codeop](https://github.com/python/cpython/blob/6da1a2e993c955aa69158871b8c8792cef3094c3/Lib/codeop.py#L60) module does this and the function does real work whereas here we are just reading an attribute. I'll also say that the issue's reproducer is IMO a contrived case.

> I like the thread local.

I can do that if you want but IMO it is too much complexity for no gain. I worry we are spending too much time on this with no gain or real issue.

Reply:
The codeop module should also use a threadsafe approach, like a thread local

Reply:
> The codeop module should also use a threadsafe approach, like a thread local

You seem to feel strongly about this, I guess you are interested in working on it? If so you can create a new PR and I'll close this.

Reply:
Honestly I'm shocked that the warnings module is not thread-safe (I haven't looked at it in many years). I feel that maybe we should make it a priority to make that at least thread-safe (possibly using contextvars to make it task-safe). In the meantime I'll approve and merge this version but I will leave the issue open so Thomas can submit a PR using a thread-local if he cares enough (it seems simple enough to me TBH).

Reply:
Making warnings thread-safe SGTM, it would solve all cases and avoid workarounds like this. but would probably require a PEP I believe since the change will be backwards incompatible.